src/transformers/models/blenderbot/modeling_blenderbot.py

```diff
@@ -252,17 +254,30 @@ def forward(
         return attn_output, attn_weights_reshaped, past_key_value
 
 
-BLENDERBOT_ATTENTION_CLASSES = {"default": BlenderbotAttention}
+BLENDERBOT_ATTENTION_CLASSES = {"eager": BlenderbotAttention}
+
+
+# Copied from transformers.models.bart.modeling_bart.BartAttentionType with Bart->Blenderbot
+class BlenderbotAttentionType(str, Enum):
+    eager = "eager"
+    sdpa = "sdpa"
+    flash_attention_2 = "flash_attention_2"
 
 
 # Copied from transformers.models.mbart.modeling_mbart.MBartEncoderLayer with MBart->Blenderbot, MBART->BLENDERBOT
 class BlenderbotEncoderLayer(nn.Module):
     def __init__(self, config: BlenderbotConfig):
         super().__init__()
         self.embed_dim = config.d_model
-        attn_type = "flash_attention_2" if getattr(config, "_flash_attn_2_enabled", False) else "default"
 
-        self.self_attn = BLENDERBOT_ATTENTION_CLASSES[attn_type](
+        if getattr(config, "_flash_attn_2_enabled", False):
```

Comment:
Really think we should follow what is proposed by @lysandrejik [here](https://github.com/huggingface/transformers/pull/27203#discussion_r1379151606): this way we ensure that we standardize the attention type at a higher level, and the attention classes should be inferred from this with a mapping.

Reply:
Thank you, I missed this comment. It makes a lot of sense. Will think about it.

However @patrickvonplaten's comment `We should then probably also allow users to change it even after the model was loaded` would mean that we would need implement a way to overload e.g. `XXXFlashAttention` by `XXXSDPAAttention` or `XXXAttention` after the model has been loaded? Alternatively, one could use an attribute `attention_type` (or rather, `attention_implementation`) in the attention layer to e.g. call `super().forward`? Looks quite involved.

Reply:
Mmmm good point. Changing the type should either be supported or restricted. I think we should not support that, will see with @patrickvonplaten as this can be done in pure torch, and we would have to support all kinds of deep speed, ddp, accelerate device etc etc dtype etc

Reply:
a) I think having a property on `PretrainedConfig` as shown here: https://github.com/huggingface/transformers/pull/27203#discussion_r1379151606 is a nice idea, but IMO can also easily be done in a follow-up PR. I don't think this needs to be done in this PR.

b) We should definitely support changing between different attention types after the model is loaded as it opens some important use cases:
- Torch.compile works better with SDPA then Flash Attention 2, it'd be nice for the user to be able to compare the two methods "on-the-fly" depending on what hardware is used without having to re-load all the weights
- During training one might want to switch between flash decoding / flash attention 2 during eval and SDPA for training (as it can be more memory efficient)
- For testing & reproducibility it's important to switch between original attention and the other ones, e.g. for such issues: https://github.com/huggingface/transformers/issues/27050#issuecomment-1798462164
- The only difference between the attention functions is the function itself, the weights stay exactly the same, so I don't see a reason why we should allow switching on the fly.

I don't think it's super important to support it right away, but it'll be good to keep it in mind when designing an API. But here again I don't think https://github.com/huggingface/transformers/pull/27203#discussion_r1379151606 is a bad design for this as we'll make the attn_type a property so we can easily adapt the property to become dynamic once we add "on-the-fly" switching

Reply:
Got it, then I think (after offline talk with @LysandreJik ) the simplest is to allow forward remapping (because the hooks are in the __call__ not the `forward`) but the attention class shown will not be the same and we would need to enforce same forward signature (which is already pretty much the case). 

But yeah to keep in ming!

Reply:
forward remapping looks like a great solution if it is tested properly, the main corner cases to test is with models loaded with accelerate and check if the hooks are correctly there. I am happy to help on that in a follow up PR as patrick suggested above!

Reply:
What hooks do we have in the attention forward here? Can't we transfer them as well? Forward mapping is pretty hacky no?

Reply:
mainly there are hooks to set the output to the correct device. Note there are also hooks for CPU / disk offloaded which are a bit technical to deal with (e.g. weights being on meta device for these cases) - cc @SunMarc in case I missed anything

Reply:
Technically we can detach the old hooks and attach them to the new module as well